File: frontend/src/api/v3/graph.ts
/**
 * Graph Analysis API Client
 * MDP Platform V3.1 - Graph Analysis Module
 *
 * Provides functions for graph expansion, shortest path, and statistics.
 */

import axiosInstance from "../axios";

// ==========================================
// Types & Interfaces
// ==========================================

export interface IGraphNode {
  id: string;
  type: string;
  position: { x: number; y: number };
  data: {
    label: string;
    object_type: string;
    icon: string;
    color: string;
    type_label: string;
    is_seed?: boolean;
  };
}

export interface IGraphEdge {
  id: string;
  source: string;
  target: string;
  type: string;
  animated?: boolean;
  style?: Record<string, unknown>;
  label?: string;
  data?: {
    role?: string;
    properties?: Record<string, unknown>;
    valid_start?: string;
    valid_end?: string;
  };
}

export interface IGraphDTO {
  nodes: IGraphNode[];
  edges: IGraphEdge[];
}

export interface IExpandOptions {
  semantic?: boolean;
  limit?: number;
  depth?: number;
  link_types?: string[];
}

export interface IExpandRequest {
  seed_ids: string[];
  options?: IExpandOptions;
}

export interface IShortestPathRequest {
  source: string;
  target: string;
  max_depth?: number;
}

export interface IShortestPathResponse {
  found: boolean;
  path: string[];
  edges: IGraphEdge[];
  distance: number;
}

export interface ILinkTypeStat {
  source: string;
  target: string;
  count: number;
}

export interface IGraphStats {
  total_links: number;
  unique_nodes: number;
  link_types: ILinkTypeStat[];
}

export interface INodeTypeConfig {
  type: string;
  icon: string;
  color: string;
  label: string;
}

// ==========================================
// API Functions
// ==========================================

/**
 * Expand graph from seed nodes
 */
export async function expandGraph(
  seedIds: string[],
  options?: IExpandOptions
): Promise<IGraphDTO> {
  const response = await axiosInstance.post<IGraphDTO>("/../v3/graph/expand", {
    seed_ids: seedIds,
    options,
  });
  return response.data;
}

/**
 * Find shortest path between two nodes
 */
export async function findShortestPath(
  source: string,
  target: string,
  maxDepth?: number
): Promise<IShortestPathResponse> {
  const response = await axiosInstance.post<IShortestPathResponse>(
    "/../v3/graph/shortest-path",
    {
      source,
      target,
      max_depth: maxDepth || 10,
    }
  );
  return response.data;
}

/**
 * Get graph statistics
 */
export async function getGraphStats(): Promise<IGraphStats> {
  const response = await axiosInstance.get<IGraphStats>("/../v3/graph/stats");
  return response.data;
}

/**
 * Get neighbors of a specific node
 */
export async function getNodeNeighbors(
  nodeId: string,
  limit?: number,
  includeSemantic?: boolean
): Promise<IGraphDTO> {
  const params = new URLSearchParams();
  if (limit) params.append("limit", limit.toString());
  if (includeSemantic) params.append("include_semantic", "true");

  const response = await axiosInstance.get<IGraphDTO>(
    `/../v3/graph/node/${nodeId}?${params.toString()}`
  );
  return response.data;
}

/**
 * Get available node types and their visual configuration
 */
export async function getNodeTypes(): Promise<{ types: INodeTypeConfig[] }> {
  const response = await axiosInstance.get<{ types: INodeTypeConfig[] }>(
    "/../v3/graph/types"
  );
  return response.data;
}

// ==========================================
// Helper Functions
// ==========================================

/**
 * Get color for a node type
 */
export function getNodeColor(objectType: string): string {
  const colorMap: Record<string, string> = {
    target: "#ff4d4f",
    vessel: "#1890ff",
    aircraft: "#52c41a",
    mission: "#722ed1",
    sensor: "#faad14",
    intel_report: "#13c2c2",
    port: "#eb2f96",
    command_unit: "#fa541c",
    recon_image: "#2f54eb",
    source: "#a0d911",
  };
  return colorMap[objectType] || "#8c8c8c";
}

/**
 * Get icon for a node type
 */
export function getNodeIcon(objectType: string): string {
  const iconMap: Record<string, string> = {
    target: "🎯",
    vessel: "🚢",
    aircraft: "✈️",
    mission: "📋",
    sensor: "📡",
    intel_report: "📄",
    port: "⚓",
    command_unit: "🎖️",
    recon_image: "📷",
    source: "🔍",
  };
  return iconMap[objectType] || "📦";
}

/**
 * Get Chinese label for a node type
 */
export function getNodeTypeLabel(objectType: string): string {
  const labelMap: Record<string, string> = {
    target: "目标",
    vessel: "舰船",
    aircraft: "飞机",
    mission: "任务",
    sensor: "传感器",
    intel_report: "情报",
    port: "港口",
    command_unit: "指挥单位",
    recon_image: "侦察图像",
    source: "情报源",
  };
  return labelMap[objectType] || "对象";
}

/**
 * Convert backend graph data to React Flow format
 */
export function toReactFlowFormat(graphDTO: IGraphDTO): {
  nodes: IGraphNode[];
  edges: IGraphEdge[];
} {
  // Fix node icons and labels (backend sends Unicode, we use local mapping)
  const nodes = graphDTO.nodes.map((node) => ({
    ...node,
    id: String(node.id), // UNIT 3: Ensure ID is always string
    data: {
      ...node.data,
      icon: getNodeIcon(node.data.object_type),
      type_label: getNodeTypeLabel(node.data.object_type),
    },
  }));

  // UNIT 3: Ensure edge IDs, sources, and targets are strings
  const edges = graphDTO.edges.map((edge) => ({
    ...edge,
    id: String(edge.id),
    source: String(edge.source),
    target: String(edge.target),
  }));

  // UNIT 3: Validate ID matching
  const nodeIdSet = new Set(nodes.map(n => n.id));
  const invalidEdges = edges.filter(
    e => !nodeIdSet.has(e.source) || !nodeIdSet.has(e.target)
  );
  
  if (invalidEdges.length > 0) {
    console.warn('[toReactFlowFormat] Found edges with missing nodes:', invalidEdges);
  }

  return { nodes, edges };
}

File: backend/app/api/v3/graph.py
"""
V3 Graph API - Graph Analysis Endpoints
MDP Platform V3.1 - Graph Analysis Module

Provides endpoints for:
- Node expansion (hard links + semantic links)
- Shortest path finding
- Graph statistics
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List
from pydantic import BaseModel
from loguru import logger

from app.services.graph_service import get_graph_service
from app.models.graph import (
    ExpandRequest, ExpandOptions,
    GraphDTO, GraphNode, GraphEdge,
    ShortestPathRequest, ShortestPathResponse
)

router = APIRouter(prefix="/graph", tags=["Graph Analysis"])


# ==========================================
# Request/Response DTOs
# ==========================================

class ExpandRequestBody(BaseModel):
    """Request body for expand endpoint."""
    seed_ids: List[str]
    options: Optional[ExpandOptions] = None


class GraphStatsResponse(BaseModel):
    """Response for graph statistics."""
    total_links: int
    unique_nodes: int
    link_types: List[dict]


# ==========================================
# Endpoints
# ==========================================

@router.post("/expand", response_model=GraphDTO)
async def expand_graph(request: ExpandRequestBody):
    """
    Expand graph from seed nodes.
    
    Supports two modes:
    - Hard Links: Database-stored relationships
    - Semantic Links: Vector similarity (when options.semantic=true)
    
    Request body:
    ```json
    {
      "seed_ids": ["tgt-001"],
      "options": {
        "semantic": false,
        "limit": 50,
        "depth": 1
      }
    }
    ```
    
    Returns React Flow compatible nodes and edges.
    """
    logger.info(f"Graph expand request: seeds={request.seed_ids}")
    
    if not request.seed_ids:
        raise HTTPException(status_code=400, detail="seed_ids cannot be empty")
    
    if len(request.seed_ids) > 10:
        raise HTTPException(status_code=400, detail="Maximum 10 seed nodes allowed")
    
    service = get_graph_service()
    result = service.expand_node(request.seed_ids, request.options)
    
    return result


@router.post("/shortest-path", response_model=ShortestPathResponse)
async def find_shortest_path(request: ShortestPathRequest):
    """
    Find shortest path between two nodes.
    
    Uses BFS to find the shortest path through the graph.
    
    Request body:
    ```json
    {
      "source": "tgt-001",
      "target": "mission-001",
      "max_depth": 10
    }
    ```
    """
    logger.info(f"Shortest path request: {request.source} -> {request.target}")
    
    if request.source == request.target:
        return ShortestPathResponse(found=True, path=[request.source], distance=0)
    
    service = get_graph_service()
    result = service.find_shortest_path(
        request.source,
        request.target,
        request.max_depth
    )
    
    return result


@router.get("/stats", response_model=GraphStatsResponse)
async def get_graph_stats():
    """
    Get graph statistics.
    
    Returns total links, unique nodes, and link type distribution.
    """
    service = get_graph_service()
    stats = service.get_graph_stats()
    
    return GraphStatsResponse(**stats)


@router.get("/node/{node_id}")
async def get_node_neighbors(
    node_id: str,
    limit: int = Query(default=20, ge=1, le=100),
    include_semantic: bool = Query(default=False)
):
    """
    Get neighbors of a specific node.
    
    Convenience endpoint for single-node expansion.
    """
    service = get_graph_service()
    options = ExpandOptions(semantic=include_semantic, limit=limit)
    result = service.expand_node([node_id], options)
    
    return result


@router.get("/types")
async def get_node_types():
    """
    Get available node types and their visual configuration.
    """
    from app.services.graph_service import NODE_TYPE_CONFIG
    
    return {
        "types": [
            {
                "type": key,
                "icon": config["icon"],
                "color": config["color"],
                "label": config["label"]
            }
            for key, config in NODE_TYPE_CONFIG.items()
        ]
    }
